### PR TITLE
Segment labels: Docker

### PR DESCRIPTION
**Pull request by ldez**
### What does this PR do?

"Service labels" become "segment labels".

Full rewrite of the template management in Docker.

### Motivation

- have a shared approach for segment labels between provider
- fix some bugs

### More

- [x] Added/updated tests
- [x] Added/updated documentation

### Additional Notes

It's the first step, next: Marathon and Rancher.

### CODE REVIEW

File: provider/label/label.go [hunk 1]
```diff
@@ -424,10 +426,85 @@ func SplitAndTrimString(base string, sep string) []string {
...
+		if matches == nil {
+			// Classic labels
+			allLabels[""][name] = value
+		} else {
+			// segments labels
+			var segmentName string
+			var propertyName string
+			for i, name := range SegmentPropertiesRegexp.SubexpNames() {
+				if i != 0 {
+					if name == "service_name" {
```
Comment (emilevauge) [inline provider/label/label.go]: s/service_name/segment_name

File: docs/configuration/backends/marathon.md [hunk 2]
```diff
@@ -235,63 +235,65 @@ The following labels can be defined on Marathon applications. They adjust the be
 | `traefik.frontend.headers.referrerPolicy=VALUE`          | Adds referrer policy  header.                                                                                                                                                                       |
 | `traefik.frontend.headers.isDevelopment=false`           | This will cause the `AllowedHosts`, `SSLRedirect`, and `STSSeconds`/`STSIncludeSubdomains` options to be ignored during development.<br>When deploying to production, be sure to set this to false. |
 
-### Service Level
+### On application with multiple ports (segment label)
 
-For applications that expose multiple ports, specific labels can be used to extract one frontend/backend configuration pair per port. Each such pair is called a _service_. The (freely choosable) name of the service is an integral part of the service label name.
+For applications that expose multiple ports, specific labels can be used to extract one frontend/backend configuration pair per port.
+Each such pair is called a _segment_.
+The (freely choosable) name of the segment is an integral part of the segment label name.
```
[the remark this answers is not in the document]
Comment (ldez) [reply docs/configuration/backends/marathon.md]: For now behavior, between Docker and Marathon on segment labels are not consistent (ex: Docker use container labels as default and not Marathon).

I will update the documentation when I will rewrite the Marathon provider.

File: provider/label/label.go [hunk 3]
```diff
@@ -257,40 +257,42 @@ func FindServiceSubmatch(name string) []string {
...
+		matches := FindSegmentSubmatch(name)
 		if matches == nil {
 			continue
 		}
 
-		var serviceName string
+		var segmentName string
 		var propertyName string
-		for i, name := range ServicesPropertiesRegexp.SubexpNames() {
+		for i, name := range SegmentPropertiesRegexp.SubexpNames() {
```
Comment (emilevauge) [inline provider/label/label.go]: I would add a comment here to explain why `if i != 0 {`

File: provider/label/label.go [hunk 4]
```diff
@@ -424,10 +426,85 @@ func SplitAndTrimString(base string, sep string) []string {
...
+		matches := FindSegmentSubmatch(name)
+		if matches == nil {
+			// Classic labels
+			allLabels[""][name] = value
+		} else {
+			// segments labels
+			var segmentName string
+			var propertyName string
+			for i, name := range SegmentPropertiesRegexp.SubexpNames() {
+				if i != 0 {
```
Comment (emilevauge) [inline provider/label/label.go]: I would add a comment here to explain why if `i != 0 {`

File: provider/label/label.go [hunk 5]
```diff
@@ -424,10 +426,85 @@ func SplitAndTrimString(base string, sep string) []string {
...
+	}
+	log.Debug(originLabels, allLabels)
+
+	allLabels.mergeDefault()
+
+	return allLabels
+}
+
+func (s SegmentProperties) mergeDefault() {
+	if defaultLabels, okDefault := s[""]; okDefault {
```
Comment (emilevauge) [inline provider/label/label.go]: A comment here would be welcome

**Pull request merged**